architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,793 @@
+
+## Assumptions
+
+1. **User permissions**: Users running the upgrade command have enough permissions on both the Kubernetes cluster and the Radius installation.
+
+2. **Resource requirements**: The Kubernetes cluster has sufficient compute resources (CPU, memory) to run both the existing and new version components during the rolling upgrade process.
+
+3. **CLI version compatibility**: User has a Radius CLI version that includes the `rad upgrade kubernetes` feature. While older CLIs can't perform upgrades, newer CLIs maintain backward compatibility with older control planes.
+
+   Example: If you have CLI v0.42 and Control Plane v0.42:
+
+   - You CAN upgrade to Control Plane v0.44 using the v0.42 CLI (if v0.42 CLI includes the upgrade feature)
+   - After upgrading, your CLI (v0.42) will still work with basic operations against your v0.44 Control Plane
+   - However, you won't be able to use any new v0.44 features from your v0.42 CLI
+   - For full functionality, you should upgrade your CLI to match the Control Plane version after the upgrade completes
+
+4. **Network connectivity**: The upgrade process requires internet connectivity to pull container images from registries and validate versions. The process is not intended to run in air-gapped environments.
+
+   - An air-gapped environment is one where systems are physically isolated from unsecured networks like the public internet.
+   - These environments are common in high-security scenarios (military, financial, healthcare, government) where external network connectivity is restricted.
+
+5. **Stable starting state**: The Radius installation being upgraded is in a healthy, stable state. Attempting to upgrade an already failing installation may lead to unpredictable results.
+
+6. **Storage availability**: The Kubernetes cluster has sufficient persistent storage capacity for backup operations during the upgrade process.
+
+## Objectives
+
+> **Issue Reference:** <https://github.com/radius-project/radius/issues/8095>
+
+### Goals
+
+- **Simplify upgrade process**: Provide a single CLI command (`rad upgrade kubernetes`) to upgrade Radius without manual reinstallation.
+- **Ensure data safety**: Implement automatic user data backups before upgrades and restore capability if failures occur.
+- **Minimize downtime**: Use rolling upgrades where possible to keep Radius control plane available during the upgrade process.
+- **Preserve application continuity**: Ensure that user applications continue running without interruption throughout the Radius upgrade process.
+
+### Non goals
+
+- **Downgrade support**: The upgrade process is designed to move forward to newer versions only. Downgrading to previous versions is not supported.
```

Comment:
one of the use cases I've heard is that an upgrade may go successfully from a Radius control plane perspective, but then subsequently result in test failures in other parts of the stack (e.g. if Radius is built into a developer platform that consists of many components downstream from Radius). how will we address or advise users for this?

Reply:
In one of the future iterations, we will add a new command to rollback to the most recent successful version of Radius. A command like `rad rollback kubernetes`. Would that answer your question?